detection_rules/generic_loader.py

```diff
@@ -0,0 +1,176 @@
+        id_map = self.id_map
+        file_map = self.file_map
+        name_map = self.name_map
+
+        assert not self.frozen, f"Unable to add item {item.name} {item.id} to a frozen collection"
+        assert item.id not in id_map, \
+            f"Rule ID {item.id} for {item.name} collides with rule {id_map.get(item.id).name}"
+        assert item.name not in name_map, \
+            f"Rule Name {item.name} for {item.id} collides with rule ID {name_map.get(item.name).id}"
+
+        if item.path is not None:
+            item_path = item.path.resolve()
+            assert item_path not in file_map, f"Item file {item_path} already loaded"
+            file_map[item_path] = item
+
+    def add_item(self, item: GenericCollectionTypes):
+        self._assert_new(item)
+        self.id_map[item.id] = item
+        self.name_map[item.name] = item
+        self.items.append(item)
+
+    def load_dict(self, obj: dict, path: Optional[Path] = None) -> GenericCollectionTypes:
+        is_exception = True if 'exceptions' in obj else False
+        contents = TOMLExceptionContents.from_dict(obj) if is_exception else TOMLActionContents.from_dict(obj)
+        item = TOMLException(path=path, contents=contents)
+        self.add_item(item)
+        return item
+
+    def load_file(self, path: Path) -> GenericCollectionTypes:
+        try:
+            path = path.resolve()
+
+            # use the default generic loader as a cache.
+            # if it already loaded the item, then we can just use it from that
+            if self.__default is not None and self is not self.__default:
+                if path in self.__default.file_map:
+                    item = self.__default.file_map[path]
+                    self.add_item(item)
+                    return item
+
+            obj = self._load_toml_file(path)
+            return self.load_dict(obj, path=path)
+        except Exception:
+            print(f"Error loading item in {path}")
+            raise
+
+    def load_files(self, paths: Iterable[Path]):
+        """Load multiple files into the collection."""
+        for path in paths:
+            self.load_file(path)
+
+    def load_directory(self, directory: Path, recursive=True, toml_filter: Optional[Callable[[dict], bool]] = None):
+        paths = self._get_paths(directory, recursive=recursive)
+        if toml_filter is not None:
+            paths = [path for path in paths if toml_filter(self._load_toml_file(path))]
+
+        self.load_files(paths)
+
+    def load_directories(self, directories: Iterable[Path], recursive=True,
+                         toml_filter: Optional[Callable[[dict], bool]] = None):
+        for path in directories:
+            self.load_directory(path, recursive=recursive, toml_filter=toml_filter)
+
+    def freeze(self):
+        """Freeze the generic collection and make it immutable going forward."""
+        self.frozen = True
+
+    @classmethod
+    def default(cls) -> 'GenericCollection':
```

Comment:
Will there be unit tests that load this Generic Collection? Just thinking about how a user will test this prior to trying to push to kibana and loading everything.

Reply:
Not built in - but if they wanted to unit test similar to how we do rules, they would just need to load and test

Reply:
you mean use the loader to "test" in terms of schema validation, but no unit testing explicitly for field validation?